pkg/solana/logpoller/log_poller.go

```diff
@@ -0,0 +1,316 @@
+			// of log corresponding filter won't be present in the db. Ensure to refilter and retry on insert error
+			for _, filter := range filtersToBackfill {
+				go lp.startFilterBackfill(ctx, filter, block.BlockNumber)
+			}
+		}
+	}
+}
+
+func (lp *LogPoller) backgroundWorkerRun() {
+	defer lp.wg.Done()
+	ctx, cancel := lp.chStop.NewCtx()
+	defer cancel()
+
+	pruneFilters := services.NewTicker(time.Minute + 618*time.Millisecond) // try to minimize collisions with one-second period
+	defer pruneFilters.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-pruneFilters.C:
+			err := lp.pruneFilters(ctx)
+			if err != nil {
+				lp.lggr.Errorw("Failed to prune filters", "err", err)
+			}
+		}
+	}
+}
+
+func (lp *LogPoller) pruneFilters(ctx context.Context) error {
+	err := lp.loadFilters(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to load filters: %w", err)
+	}
+
+	lp.filtersMutex.Lock()
+	filtersToDelete := lp.filtersToDelete
+	lp.filtersToDelete = nil
+	lp.filtersMutex.Unlock()
+
+	if len(filtersToDelete) == 0 {
+		return nil
+	}
+
+	err = lp.orm.DeleteFilters(ctx, filtersToDelete)
+	if err != nil {
+		lp.filtersMutex.Lock()
+		defer lp.filtersMutex.Unlock()
+		lp.filtersToDelete = append(lp.filtersToDelete, filtersToDelete...)
+		return fmt.Errorf("failed to delete filters: %w", err)
+	}
+
+	return nil
+}
+
+func (lp *LogPoller) startFilterBackfill(ctx context.Context, filter Filter, toBlock int64) {
+	// TODO: NONEVM-916 start backfill
+	lp.lggr.Debugw("Starting filter backfill", "filter", filter)
+}
+
+// RegisterFilter persists provided filter and ensures that any log emitted by a contract with filter.Address
+// that matches filter.EventSig signature will be captured starting from filter.StartingBlock.
+// filter.Name must be unique otherwise ErrFilterNameConflict is returned. // TODO: not sure this is a good idea. Callers are most likely going to ignore this error
+// The filter may be unregistered later by Filter.Name
+// Warnings/debug information is keyed by filter name.
+func (lp *LogPoller) RegisterFilter(ctx context.Context, filter Filter) error {
+	if len(filter.Name) == 0 {
+		return errors.New("name is required")
+	}
+
+	err := lp.loadFilters(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to load filters: %w", err)
+	}
+
+	lp.filtersMutex.Lock()
+	defer lp.filtersMutex.Unlock()
+
+	if _, ok := lp.filtersByName[filter.Name]; ok {
+		return ErrFilterNameConflict
+	}
+
+	filterID, err := lp.orm.InsertFilter(ctx, filter)
+	if err != nil {
+		return fmt.Errorf("failed to insert filter: %w", err)
+	}
+
+	filter.ID = filterID
+	lp.filtersByName[filter.Name] = filter
+	filtersByEventSig, ok := lp.filtersByAddress[filter.Address]
+	if !ok {
+		filtersByEventSig = make(map[EventSignature][]Filter)
+		lp.filtersByAddress[filter.Address] = filtersByEventSig
+	}
+
+	filtersByEventSig[filter.EventSig] = append(filtersByEventSig[filter.EventSig], filter)
+	lp.filtersToBackfill = append(lp.filtersToBackfill, filter)
```

Comment:
> (If a log matches multiple filters, then multiple copies of it will be saved.) 

That reminds me, we'll need to include an extra deduping step in the ORM when logs are fetched

Reply:
Oh, now I recall main argument for the backfill queue. 
As blocks are processed in effectively random order and Backfill operation itself is async we do not have "latest processed block" as in evm. Thus backfill should be done one every restart too. Having backfill queue allows us to do it in one place.
Also when RegisterFilter is called we might not have the `latestProcessedBlock` to start the backfill.

Reply:
I'm fine with keeping it as a queue, but I don't follow (and/or agree with?) either the premise or the conclusion of this statement:
> As blocks are processed in effectively random order random order and Backfill operation itself is async we do not have "latest processed block" as in evm.

Perhaps by "processed" you mean something else, but I thought we had all agreed that we must guarantee they are always saved in order of (block_number, log_index). In part to make sure that there is a well-defined last processed block, but mostly to avoid introducing gaps in the logs.  You pointed out that Awbrey's ingestor fetches them in random order from the rpc, which I agreed was a big problem. So we discussed the problem with Awbrey, and we agreed that they must be re-ordered somewhere.  But he made the argument that it makes more sense to do the reordering in the callback function passed to the ingestor rather than in his PR (the ingestor itself). The callback function will be a part of the PR I'm starting next and include the re-ordering so we can ensure this guarantee.

The design of LogPoller in the design doc is based around the assumption that we always have a well-defined "latest processed block". The only difference there from evm is that because we don't have a blocks table, we will get this by fetching the latest block number from the `block_number` column of the logs table instead of the `block_number` column of the blocks table. It's true that if the node crashes and the last few blocks had no logs in them, we may end up re-requesting a few empty blocks. But we shouldn't have to do anything special for that in terms of the way we're fetching logs.

A side digression on the term "backfill"... which I keep thinking we might want to avoid using on Solana due to confusion:

On evm, the distinction between a backfill and a non-backfill fetching of logs is whether we fetch logs from a range of block numbers or from a single block by block hash.  On Solana, we don't have to worry about re-orgs so there is no reason for the second method of fetching logs--we always fetch them by block range rather than by individual blocks.  However, we do have a choice of whether to iterate through specific contract addresses we're tracking and only fetch the logs from each of those or fetching all logs and then post-filtering on contract address.  Under normal operations, the decision of whether to filter by contract address or to just get everything will be based on a combination of how many blocks have been mined since the last time LogPoller ran, and how many filters are registered.  If it's a lot of filters and only a few new blocks have been added to the chain since last time we polled then it makes more sense to grab everything in a single rpc call and then filter. If on the other hand, there are only a few filters registered but we're thousands of blocks behind because the node has been down for a while, it makes more sense to first ask the rpc servers which blocks actually contain transactions associated with the addresses we're interested in and then *only* request the specific blocks and/or transactions we need instead of fetching every tx of every block.

RegisterFilter in the only case where we know ahead of time which of those methods will be best to use: since it's only for a single filter, we should always ask first about that address rather than getting all of the blocks... regardless of how far back the starting block is or how many filters have been registered.

For now, based on the way Awbrey named the methods I'm assuming when we say "backfill" on Solana we mean the case where we're fetching log history for a single filter rather than for all filters at once.  But it might be there is a better term for that, so we don't confuse it with the use of "backfill" in evm which I think means something pretty different.

Reply:
Oh, I think the reason I decided to clarify what I think we mean by "backfill" when we're using that term is because I got a little confused by this statement:
> Backfill operation itself is async

By that, I'm guessing what you mean is that RegisterFilter is async... in that it can be called at any time by any thread, while the main LogPoller loop is doing its thing. The backfill itself we could have chosen to make synchronous or async, the way I see it is... you've chosen to make it async by having RegisterFilter return immediately rather than wait for it to complete. Yes?

Reply:
I agree with your point about not wanting to lock the filters for a long time. Might have been some other workaround but your way of handling it sounds reasonable to me. I'm not sure about the issue of UI, I think most products are already doing something similar on evm (calling RegisterFilter then Replay immediately after, which blocks until everything is backfilled). AFAIK this has been working like this for a while and hasn't caused any problems. But you could be right that we should be extra careful with Solana since these operations could potentially be more time consuming than on evm.

Reply:
I was not sure if we actually decided to do the ordering. Thus, I've tried to tailor work with filters to the current Loader implementation. But it works for the "ordered" version too. 
As we both agree on keeping the queue, lets move discussion into slack or sync meeting. It would be nice to discuss how ordering will work.